code/extensions/LDAPMemberExtension.php

```diff
@@ -95,14 +136,86 @@ public function updateCMSFields(FieldList $fields)
         }
     }
 
+    public function validate(ValidationResult $validationResult)
+    {
+        if (!$this->owner->config()->create_users_in_ldap) {
+            return;
+        }
+
+        // We allow empty Username for registration purposes, as we need to
+        // create Member records with empty Username temporarily. Forms should explicitly
+        // check for Username not being empty if they require it not to be.
+        if (empty($this->owner->Username)) {
+            return;
+        }
+
+        if (!preg_match('/^[a-z0-9\.]+$/', $this->owner->Username)) {
+            $validationResult->error(
+                'Username must only contain lowercase alphanumeric characters and dots.',
+                'bad'
+            );
+            throw new ValidationException($validationResult);
+        }
+    }
+
+    /**
+     * Create the user in LDAP, provided this configuration is enabled
+     * and a username was passed to a new Member record.
+     */
+    public function onBeforeWrite()
+    {
+        $service = Injector::inst()->get('LDAPService');
+        if (!$service->enabled()) {
+            return;
+        }
+        if (!$this->owner->config()->create_users_in_ldap) {
+            return;
+        }
+        if ($this->owner->Username && !$this->owner->IsImportedFromLDAP) {
+            $service->createLDAPUser($this->owner);
+        }
+    }
+
+    /**
+     * Update the local data with LDAP, and ensure local membership is also set in
+     * LDAP too. This writes into LDAP, provided that feature is enabled.
+     */
+    public function onAfterWrite()
+    {
+        $service = Injector::inst()->get('LDAPService');
+        if (!$service->enabled()) {
+            return;
+        }
+        if (!$this->owner->config()->update_ldap_from_local) {
+            return;
+        }
+        if ($this->owner->IsImportedFromLDAP) {
+            $service->updateLDAPFromMember($this->owner);
+            $service->updateLDAPGroupsForMember($this->owner);
+        }
+    }
+
+    public function onAfterDelete() {
+        $service = Injector::inst()->get('LDAPService');
+        if (!$service->enabled()) {
+            return;
+        }
+        if (!$this->owner->config()->delete_users_in_ldap) {
+            return;
+        }
+        if ($this->owner->IsImportedFromLDAP) {
+            $service->deleteLDAPMember($this->owner);
+        }
+    }
+
     /**
      * Triggered by {@link Member::logIn()} when successfully logged in,
      * this will update the Member record from AD data.
      */
     public function memberLoggedIn()
     {
         if ($this->owner->GUID) {
```

Comment:
we are using two different variabels for updating the ldap data, `GUID` and `IsImportedFromLDAP` can we be more coherent.. if possible.